engine/runtime/src/main/scala/org/enso/interpreter/runtime/scope/LocalScope.scala

```diff
@@ -28,20 +28,35 @@ import scala.jdk.CollectionConverters._
   * @param dataflowInfo information on the dataflow analysis for this scope
   * @param flattenToParent whether or not the frame should be flattened into its
   *                        parent
-  * @param frameSlots a mapping from symbol definition identifiers to slots in
-  *                   the Enso frame
+  * @param parentFrameSlotIdxs Mapping of occurence identifiers to frame slot indexes
+  *       from the whole parent hierarchy, i.e., this parameter should contain all the
+  *       indexes for all the parents.
   */
 class LocalScope(
   final val parentScope: Option[LocalScope],
   final val aliasingGraph: AliasAnalysis.Graph,
   final val scope: AliasAnalysis.Graph.Scope,
   final val dataflowInfo: DataflowAnalysis.Metadata,
-  final val flattenToParent: Boolean                     = false,
-  final val frameSlots: mutable.Map[Graph.Id, FrameSlot] = mutable.Map()
+  final val flattenToParent: Boolean                  = false,
+  private val parentFrameSlotIdxs: Map[Graph.Id, Int] = Map()
 ) {
+  lazy val frameDescriptor: FrameDescriptor = buildFrameDescriptor()
+  private lazy val localFrameSlotIdxs: Map[Graph.Id, Int] =
+    gatherLocalFrameSlotIdxs()
 
-  /** A descriptor for this frame. */
-  val frameDescriptor: FrameDescriptor = new FrameDescriptor()
+  /** All frame slot indexes, including local and all the parents.
+    * Useful for quick searching for [[FramePointer]] of parent scopes.
+    */
+  private lazy val allFrameSlotIdxs: Map[Graph.Id, Int] =
+    parentFrameSlotIdxs ++ localFrameSlotIdxs
+
+  /** Internal slots are prepended at the beginning of every [[FrameDescriptor]].
+    * Every tuple of the list denotes frame slot kind and its name.
+    * Note that `info` for a frame slot is not used by Enso.
+    */
+  private val internalSlots: List[(FrameSlotKind, String)] = List(
```

Comment:
It feels like this is constant and should be defined in a companion object of `LocalScope`